docs/design-documents/platform/Statistics/NetworkStatistics.md

```diff
@@ -0,0 +1,186 @@
+|----------------------------------------|      `------------------------'      `--------------------'      `----'
+|+ <module>_stats_t _stats;              |                 |Uses                          |Uses
+|----------------------------------------|                 |                              |
+|+ inline void log_xx();                 |                 |                              |
+|  ...                                   |                 |                              |
+|+ virtual void read_stats(...);         |                 |                              |
+`----------------------------------------'                 |                              |
+               |Uses                                       |                              |
+               |                                           |                              |
+,---------------------------------.     ,---------------------------------.   ,-----------------------------.
+| Class: <Module>                 |     |Class: EMACInterface             |   |Class: LWIP                  |
+|#Object:<Module_name>Stats stats;|     |#Object:EMACInterfaceStats stats;|   |#Object:LWIPStackStats stats;|
+`---------------------------------'     `---------------------------------'   `-----------------------------'
+```
+
+# Detailed design
+
+## StatsList Class
+An abstract class used to create a linked list entry of every module reporting statistics data. Static API's in StatsList class can be used to read statistics data stored in linked list.
+
+#### Data Structures
+
+1. Statistic types for segregation of different modules and can also be used for filtering statistics data.
+```
+/** Stats type */
+typedef enum {
+    STATS_RTOS_GEN = 0x0,
+
+    STATS_PLATFORM_GEN = 0x20,
+
+    STATS_STORAGE_GEN = 0x40,
+
+    STATS_NW_GEN = 0x60,
+    STATS_NW_INTERFACE_EMAC,
+    STATS_NW_INTERFACE_WIFI,
+    STATS_NW_INTERFACE_MESH,
+    STATS_NW_INTERFACE_CELLULAR,
+
+    STATS_NW_STACK_LWIP,
+    STATS_NW_STACK_WIFI,
+}mbed_stats_type_t;
+```
+1. Statistic Info Structure - Structure used to get statistics data as binary blob from the StatsList class.
+
+```
+typedef struct  {
+    mbed_stats_type_t stats_type;
+    uint32_t buf_size;
+    void *buf;
+}stats_info_t;
+```
+
+#### API's
+
+1. static int get_each(stats_info_t *stats, int count);
+API to fetch all the statistics data available in the linked list. User can specify how many elements from linked list should be fetched in `count` parameter.
+
+2. static int get_each(stats_info_t *stats, mbed_stats_type_t type, int count);
+API to fetch specific `type` of statistic data available in the linked list. User can specify how many elements from linked list should be fetched in `count` parameter.
+
+## Ethernet Statistics
+
+EMAC interface class and LWIP stack together provided all information required for ethernet stattistics. `EMACInterfaceStats` and `LWIPStackStats` classes provide mechanism to log all statistics information.
```

Comment:
Why do you try to design here how drivers internally gathers the data.

We should not be dictating the API design into the drivers.